rust/tool/systemd/src/install.rs

```diff
@@ -380,89 +349,37 @@ impl Installer {
-    /// GenerationArtifacts.
-    tempdir: TempDir,
-
-    /// A mapping from target location to source.
-    files: BTreeMap<PathBuf, FileSource>,
+/// Translate an EFI path to an absolute path on the mounted ESP.
+fn resolve_efi_path(esp: &Path, efi_path: &[u8]) -> Result<PathBuf> {
+    Ok(esp.join(std::str::from_utf8(&efi_path[1..])?.replace('\\', "/")))
 }
 
-impl GenerationArtifacts {
-    fn new() -> Result<Self> {
-        Ok(Self {
-            tempdir: TempDir::new().context("Failed to create temporary directory.")?,
-            files: Default::default(),
-        })
-    }
-
-    /// Add a file to be installed.
-    ///
-    /// Adding the same file multiple times with the same source is ok
-    /// and will drop the old source.
-    fn add_file(&mut self, from: FileSource, to: &Path) {
-        if let Some(_prev_from) = self.files.insert(to.to_path_buf(), from) {
-            // Should we log something here?
-        }
-    }
-
-    /// Add source and destination of a PE file to be signed.
-    ///
-    /// Files are stored in the HashMap using their destination path as the key to ensure that the
-    /// destination paths are unique.
-    fn add_signed(&mut self, from: &Path, to: &Path) {
-        self.add_file(FileSource::SignedFile(from.to_path_buf()), to);
-    }
-
-    /// Add source and destination of an arbitrary file.
-    fn add_unsigned(&mut self, from: &Path, to: &Path) {
-        self.add_file(FileSource::UnsignedFile(from.to_path_buf()), to);
-    }
-
-    /// Install all files to the ESP.
-    fn install(&self, key_pair: &KeyPair) -> Result<()> {
-        for (to, from) in &self.files {
-            match from {
-                FileSource::SignedFile(from) => {
-                    install_signed(key_pair, from, to).with_context(|| {
-                        format!("Failed to sign and install from {from:?} to {to:?}")
-                    })?
-                }
-                FileSource::UnsignedFile(from) => install(from, to)
-                    .with_context(|| format!("Failed to install from {from:?} to {to:?}"))?,
-            }
-        }
-
-        Ok(())
+/// Compute the file name to be used for the stub of a certain generation, signed with the given key.
+fn stub_name(generation: &Generation, public_key: &Path) -> Result<PathBuf> {
+    let bootspec = &generation.spec.bootspec.bootspec;
+    let stub_inputs = [
+        // Generation numbers can be reused if the latest generation was deleted.
+        // To detect this, the stub path depends on the actual toplevel used.
+        ("toplevel", bootspec.toplevel.0.as_os_str().as_bytes()),
+        // If the key is rotated, the signed stubs must be re-generated.
+        // So we make their path depend on the public key used for signature.
+        ("public_key", &fs::read(public_key)?),
+    ];
+    let stub_input_hash = nix_base32::to_nix_base32(&Sha256::digest(
+        serde_json::to_string(&stub_inputs).unwrap(),
```

Comment:
So if you want to compute a hash from two inputs, it is not a good idea to just concatenate them before hashing. For example, input 1 being "A" and input 2 being "BC" will give the same hash as input 1 being "AB" and input 2 being "C". This can be avoided by using an injective transformation instead of concatenation, which JSON encoding is.

As noted above, I think this will probably not happen in practice here, but I prefer not to have to think about this issue, and we already depend on `serde_json` (via bootspec).